### PR TITLE
Add api test for setting expiration date for resharing to user and groups

### PR DESCRIPTION
**Pull request by dpakach**
## Description

Api acceptance tests for setting expiration dates on user and group reshares.

## Related Issue

-Tests for https://github.com/owncloud/enterprise/issues/2001

## How Has This Been Tested?

locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 1]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
       | ocs_api_version | ocs_status_code | http_status_code |
       | 1               | 403             | 200              |
       | 2               | 403             | 403              |
+
+  Scenario Outline: User should be able to set expiration while resharing a file
+    Given using OCS API version "2"
```
Comment (HariBhandari07) [inline tests/acceptance/features/apiShareReshare/reShare.feature]: This line is redundant

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 2]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
...
+      | expireDate  | +3 days       |
+    And the information of the last share of user "user1" should include
+      | expiration  | +3 days    |
+    Examples:
+      | ocs_api_version |
+      | 1               |
+      | 2               |
+
+  Scenario Outline: User should be able to set expiration while resharing a file with group
+    Given using OCS API version "2"
```
Comment (HariBhandari07) [inline tests/acceptance/features/apiShareReshare/reShare.feature]: same as above

**Review by haribhandari07 — DISMISSED** [review]

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 3]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
...
+    Given using OCS API version "<ocs_api_version>"
+    And user "user2" has been created with default attributes and without skeleton files
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    And user "user1" has created a share with settings
+      | path        | textfile0.txt |
+      | shareType   | user          |
+      | permissions | change        |
+      | shareWith   | user2         |
+      | expireDate  | +3 days       |
+    And the information of the last share of user "user1" should include
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]:
```suggestion
    Then the information of the last share of user "user1" should include
```

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 4]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
...
+    Given using OCS API version "<ocs_api_version>"
+    And group "grp1" has been created
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    And user "user1" has created a share with settings
+      | path        | textfile0.txt |
+      | shareType   | group         |
+      | permissions | change        |
+      | shareWith   | grp1          |
+      | expireDate  | +3 days       |
+    And the information of the last share of user "user1" should include
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]:
```suggestion
    Then the information of the last share of user "user1" should include
```

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 5]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
...
+    And parameter "shareapi_enforce_expire_date_user_share" of app "core" has been set to "<enforce-expire-date>"
+    And parameter "shareapi_expire_after_n_days_user_share" of app "core" has been set to "30"
+    And user "user2" has been created with default attributes and without skeleton files
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    And user "user1" has created a share with settings
+      | path        | textfile0.txt |
+      | shareType   | user          |
+      | permissions | change        |
+      | shareWith   | user2         |
+    And the information of the last share of user "user1" should include
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]:
```suggestion
    Then the information of the last share of user "user1" should include
```

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 6]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
...
+    And parameter "shareapi_expire_after_n_days_group_share" of app "core" has been set to "30"
+    And group "grp1" has been created
+    And user "user1" has been added to group "grp1"
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    And user "user1" has created a share with settings
+      | path        | textfile0.txt |
+      | shareType   | group         |
+      | permissions | change        |
+      | shareWith   | grp1          |
+    And the information of the last share of user "user1" should include
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]:
```suggestion
    Then the information of the last share of user "user1" should include
```

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 7]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
       | ocs_api_version | ocs_status_code | http_status_code |
       | 1               | 403             | 200              |
       | 2               | 403             | 403              |
+
+  Scenario Outline: User should be able to set expiration while resharing a file
+    Given using OCS API version "2"
+    Given using OCS API version "<ocs_api_version>"
+    And user "user2" has been created with default attributes and without skeleton files
```
Comment (HariBhandari07) [inline tests/acceptance/features/apiShareReshare/reShare.feature]: I don't think creation of user2 serves any purpose here.

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 8]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
...
+    And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "<default-expire-date>"
+    And parameter "shareapi_enforce_expire_date_user_share" of app "core" has been set to "<enforce-expire-date>"
+    And user "user2" has been created with default attributes and without skeleton files
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    And user "user1" has created a share with settings
+      | path        | textfile0.txt |
+      | shareType   | user          |
+      | permissions | change        |
+      | shareWith   | user2         |
+    And the information of the last share of user "user1" should include
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]:
```suggestion
    Then the information of the last share of user "user1" should include
```

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 9]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
...
+    And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "<default-expire-date>"
+    And parameter "shareapi_enforce_expire_date_group_share" of app "core" has been set to "<enforce-expire-date>"
+    And group "grp2" has been created
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    And user "user1" has created a share with settings
+      | path        | textfile0.txt  |
+      | shareType   | group          |
+      | permissions | change         |
+      | shareWith   | grp2           |
+    And the information of the last share of user "user1" should include
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]:
```suggestion
    Then the information of the last share of user "user1" should include
```

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 10]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
...
+    And parameter "shareapi_expire_after_n_days_user_share" of app "core" has been set to "30"
+    And user "user2" has been created with default attributes and without skeleton files
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    And user "user1" has created a share with settings
+      | path        | textfile0.txt |
+      | shareType   | user          |
+      | permissions | change        |
+      | shareWith   | user2         |
+      | expireDate  | +20 days      |
+    And the information of the last share of user "user1" should include
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]:
```suggestion
    Then the information of the last share of user "user1" should include
```

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 11]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
...
+    And parameter "shareapi_expire_after_n_days_group_share" of app "core" has been set to "30"
+    And group "group2" has been created with default attributes
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    And user "user1" has created a share with settings
+      | path        | textfile0.txt |
+      | shareType   | group          |
+      | permissions | change        |
+      | shareWith   | group2         |
+      | expireDate  | +20 days      |
+    And the information of the last share of user "user1" should include
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]:
```suggestion
    Then the information of the last share of user "user1" should include
```

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 12]
```diff
@@ -530,3 +530,166 @@ Feature: sharing
...
+      | 2               | no                  | yes                 |
+
+  Scenario Outline: Default expiration date for resharing to a group using the sharing API with default expire date set and specify expire date in share
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "<default-expire-date>"
+    And parameter "shareapi_enforce_expire_date_group_share" of app "core" has been set to "<enforce-expire-date>"
+    And parameter "shareapi_expire_after_n_days_group_share" of app "core" has been set to "30"
+    And group "group2" has been created with default attributes
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    And user "user1" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]: Can we make this a when step?
(in all the new scenarios)

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 13]
```diff
@@ -530,3 +530,188 @@ Feature: sharing
       | ocs_api_version | ocs_status_code | http_status_code |
       | 1               | 403             | 200              |
       | 2               | 403             | 403              |
+
+  @skipOnOcV10.3
+  Scenario Outline: User should be able to set expiration while resharing a file
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]:
```suggestion
  Scenario Outline: User should not be able to set expiration while resharing a file when expiration date is not enabled
```
(the test is expecting OCS 404 - so is this scenario meant to test when the option is not enabled? or did you mean to enable the expiry date feature and then the status should be 2nn)

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 14]
```diff
@@ -530,3 +530,188 @@ Feature: sharing
...
+    And the OCS status code should be "404"
+    And the information of the last share of user "user1" should include
+      | expiration  | +3 days    |
+    Examples:
+      | ocs_api_version | http_status_code |
+      | 1               | 100              |
+      | 2               | 200              |
+
+  @skipOnOcV10.3
+  Scenario Outline: User should be able to set expiration while resharing a file with group
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]:
```suggestion
  Scenario Outline: User should not be able to set expiration while resharing a file with group when expiration is not enabled
```
(the test is expecting OCS 404 - so is this scenario meant to test when the option is not enabled? or did you mean to enable the expiry date feature and then the status should be 2nn)

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 15]
```diff
@@ -530,3 +530,188 @@ Feature: sharing
...
+    And parameter "shareapi_expire_after_n_days_user_share" of app "core" has been set to "30"
+    And user "user2" has been created with default attributes and without skeleton files
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    When user "user1" creates a share using the sharing API with settings
+      | path        | textfile0.txt |
+      | shareType   | user          |
+      | permissions | change        |
+      | shareWith   | user2         |
+    Then the HTTP status code should be "<http_status_code>"
+    And the OCS status code should be "404"
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]: why 404?

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 16]
```diff
@@ -530,3 +530,188 @@ Feature: sharing
...
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "<default-expire-date>"
+    And parameter "shareapi_enforce_expire_date_group_share" of app "core" has been set to "<enforce-expire-date>"
+    And parameter "shareapi_expire_after_n_days_group_share" of app "core" has been set to "30"
+    And group "grp1" has been created
+    And user "user1" has been added to group "grp1"
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    Then the HTTP status code should be "<http_status_code>"
+    And the OCS status code should be "404"
+    When user "user1" creates a share using the sharing API with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]: These lines should be after the When step
Why 404?

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 17]
```diff
@@ -530,3 +530,188 @@ Feature: sharing
...
+    And parameter "shareapi_enforce_expire_date_user_share" of app "core" has been set to "<enforce-expire-date>"
+    And user "user2" has been created with default attributes and without skeleton files
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    When user "user1" creates a share using the sharing API with settings
+      | path        | textfile0.txt |
+      | shareType   | user          |
+      | permissions | change        |
+      | shareWith   | user2         |
+    Then the HTTP status code should be "<http_status_code>"
+    And the OCS status code should be "404"
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]: 404?

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 18]
```diff
@@ -530,3 +530,188 @@ Feature: sharing
...
+    And parameter "shareapi_enforce_expire_date_group_share" of app "core" has been set to "<enforce-expire-date>"
+    And group "grp2" has been created
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    When user "user1" creates a share using the sharing API with settings
+      | path        | textfile0.txt  |
+      | shareType   | group          |
+      | permissions | change         |
+      | shareWith   | grp2           |
+    Then the HTTP status code should be "<http_status_code>"
+    And the OCS status code should be "404"
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]: 404?

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 19]
```diff
@@ -530,3 +530,188 @@ Feature: sharing
...
+    And user "user2" has been created with default attributes and without skeleton files
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    When user "user1" creates a share using the sharing API with settings
+      | path        | textfile0.txt |
+      | shareType   | user          |
+      | permissions | change        |
+      | shareWith   | user2         |
+      | expireDate  | +20 days      |
+    Then the HTTP status code should be "<http_status_code>"
+    And the OCS status code should be "404"
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]: 404?

File: tests/acceptance/features/apiShareReshare/reShare.feature [hunk 20]
```diff
@@ -530,3 +530,188 @@ Feature: sharing
...
+    And group "group2" has been created with default attributes
+    And user "user0" has shared file "/textfile0.txt" with user "user1" with permissions "read,update,share"
+    When user "user1" creates a share using the sharing API with settings
+      | path        | textfile0.txt |
+      | shareType   | group         |
+      | permissions | change        |
+      | shareWith   | group2        |
+      | expireDate  | +20 days      |
+    Then the HTTP status code should be "<http_status_code>"
+    And the OCS status code should be "404"
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshare/reShare.feature]: 404?

**Review by phil-davis — COMMENTED** [review]
Quick cut-paste at the end of the doesn't work so well ;)

**Review by phil-davis — APPROVED** [review]

**Pull request merged**